File: backend/app/schemas/declared.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, field_serializer


class DeclaredAccountResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    account_code: str
    account_name: str
    account_type: str | None
    account_nature: str | None
    account_level: int | None
    parent_account_code: str | None
    account_order: int | None
    declared_reference_code: str | None
    official_description: str | None
    official_reference_status: str | None
    methodology_rule_applied: str | None
    methodology_rule_status: str | None
    purpose: str
    treatment: str | None
    base_value: Decimal
    considered_value: Decimal
    final_status: str
    observation: str | None
    recommended_action: str | None
    methodology_version_id: str

    @field_serializer("base_value", "considered_value")
    def serialize_decimal(self, value: Decimal) -> str:
        return format(value, "f")


class DeclaredAccountsResponse(BaseModel):
    analysis_id: str
    year: int
    accounts: list[DeclaredAccountResponse]


class DeclaredBalanceRowResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    aggregation_code: str
    aggregation_code_type: str
    aggregation_level: int
    parent_aggregation_code: str | None
    balance_group: str
    description: str
    initial_amount: Decimal
    initial_debit_credit_indicator: str
    final_amount: Decimal
    final_debit_credit_indicator: str
    explanatory_note_reference: str | None
    line_number: int
    structural_status: str
    reconciliation_status: str | None
    reconciled_amount: Decimal | None
    difference: Decimal | None
    component_count: int
    children: list["DeclaredBalanceRowResponse"]

    @field_serializer(
        "initial_amount",
        "final_amount",
        "reconciled_amount",
        "difference",
    )
    def serialize_money(self, value: Decimal | None) -> str | None:
        return format(value, "f") if value is not None else None


class DeclaredBalanceResponse(BaseModel):
    analysis_id: str
    year: int
    balance_status: str
    is_blocking: bool
    j005_period_start: date | None
    j005_period_end: date | None
    assets_final_amount: Decimal | None
    liabilities_and_equity_final_amount: Decimal | None
    difference: Decimal | None
    rows: list[DeclaredBalanceRowResponse]
    limitations: list[str]

    @field_serializer(
        "assets_final_amount",
        "liabilities_and_equity_final_amount",
        "difference",
    )
    def serialize_money(self, value: Decimal | None) -> str | None:
        return format(value, "f") if value is not None else None


class DeclaredBalanceComponentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    account_code: str
    account_name: str
    cost_center_code: str | None
    final_amount: Decimal | None
    final_debit_credit_indicator: str | None
    signed_final_amount: Decimal | None
    i052_line_number: int
    i155_line_number: int | None

    @field_serializer("final_amount", "signed_final_amount")
    def serialize_money(self, value: Decimal | None) -> str | None:
        return format(value, "f") if value is not None else None


class DeclaredBalanceComponentsResponse(BaseModel):
    analysis_id: str
    year: int
    aggregation_code: str
    rows: list[DeclaredBalanceComponentResponse]


class DeclaredLayerSummaryResponse(BaseModel):
    analysis_id: str
    year: int
    total_accounts: int
    status_counts: dict[str, int]
    methodology_version_id: str | None


class DeclaredRunResponse(BaseModel):
    analysis_id: str
    year: int
    status: str
    snapshots_created: int
    status_counts: dict[str, int]


class ApiErrorResponse(BaseModel):
    error_code: str
    message: str

File: frontend/src/api/declared.ts
export type DeclaredLayerSummary = {
  analysis_id: string;
  year: number;
  total_accounts: number;
  methodology_version_id: string | null;
  status_counts: Record<string, number>;
};

export type DeclaredAccount = {
  account_code: string;
  account_name: string;
  account_type: string | null;
  account_nature: string | null;
  account_level: number | null;
  parent_account_code: string | null;
  account_order: number | null;
  declared_reference_code: string | null;
  official_description: string | null;
  official_reference_status: string | null;
  methodology_rule_applied: string | null;
  methodology_rule_status: string | null;
  purpose: string | null;
  plra_category: string | null;
  fco_category: string | null;
  capag_category: string | null;
  flow_nature: string | null;
  treatment: string | null;
  base_value: string;
  considered_value: string;
  final_status: string;
  observation: string | null;
  recommended_action: string | null;
  methodology_version_id: string;
};

export type DeclaredAccountsResponse = {
  analysis_id: string;
  year: number;
  accounts: DeclaredAccount[];
};

export type DeclaredBalanceStatus =
  | "VALIDO"
  | "DIVERGENTE"
  | "OBRIGATORIO_AUSENTE"
  | "ESTRUTURA_INVALIDA"
  | "NAO_OBRIGATORIO";

export type DeclaredBalanceLineStatus =
  | "CONCILIADA"
  | "DIVERGENTE"
  | "SEM_I052"
  | "SEM_SALDO_I155";

export type DeclaredBalanceRow = {
  aggregation_code: string;
  aggregation_code_type: "T" | "D";
  aggregation_level: number;
  parent_aggregation_code: string | null;
  balance_group: "A" | "P";
  description: string;
  initial_amount: string;
  initial_debit_credit_indicator: string;
  final_amount: string;
  final_debit_credit_indicator: string;
  explanatory_note_reference: string | null;
  line_number: number;
  structural_status: "VALIDA" | "INVALIDA";
  reconciliation_status: DeclaredBalanceLineStatus | null;
  reconciled_amount: string | null;
  difference: string | null;
  component_count: number;
  children: DeclaredBalanceRow[];
};

export type DeclaredBalanceResponse = {
  analysis_id: string;
  year: number;
  balance_status: DeclaredBalanceStatus;
  is_blocking: boolean;
  j005_period_start: string | null;
  j005_period_end: string | null;
  assets_final_amount: string | null;
  liabilities_and_equity_final_amount: string | null;
  difference: string | null;
  rows: DeclaredBalanceRow[];
  limitations: string[];
};

export type DeclaredBalanceComponent = {
  account_code: string;
  account_name: string;
  cost_center_code: string | null;
  final_amount: string | null;
  final_debit_credit_indicator: string | null;
  signed_final_amount: string | null;
  i052_line_number: number;
  i155_line_number: number | null;
};

export type DeclaredBalanceComponentsResponse = {
  analysis_id: string;
  year: number;
  aggregation_code: string;
  rows: DeclaredBalanceComponent[];
};

export type EcdImportResponse = {
  analysis_id: string;
  company_id: string;
  ecd_file_id: string;
  year: number;
  methodology_version_id: string;
  status: string;
  parser_version: string;
  balance_preparation_status: string;
  reprocessed: boolean;
};

export type ExistingEcdImport = EcdImportResponse & {
  original_filename: string;
  content_hash: string;
  layout: string;
  period_start: string;
  period_end: string;
  imported_at: string;
};

export type EcdImportListResponse = {
  imports: ExistingEcdImport[];
};

export type EcdImportDeleteResponse = {
  ecd_file_id: string;
  analysis_id: string;
  deleted: boolean;
};

export type DeclaredRunResponse = {
  analysis_id: string;
  year: number;
  status: string;
  snapshots_created: number;
  status_counts: Record<string, number>;
};

export class EcdImportConflictError extends Error {
  existingImport: ExistingEcdImport;

  constructor(message: string, existingImport: ExistingEcdImport) {
    super(message);
    this.name = "EcdImportConflictError";
    this.existingImport = existingImport;
  }
}

async function parseApiError(response: Response): Promise<Error> {
  try {
    const body = (await response.json()) as {
      detail?: {
        error_code?: string;
        existing_import?: ExistingEcdImport;
        message?: string;
      };
      message?: string;
    };
    const message = body.detail?.message ?? body.message ?? `HTTP ${response.status}`;

    if (
      response.status === 409 &&
      body.detail?.error_code === "ECD_ALREADY_IMPORTED" &&
      body.detail.existing_import
    ) {
      return new EcdImportConflictError(message, body.detail.existing_import);
    }

    return new Error(message);
  } catch {
    return new Error(`HTTP ${response.status}`);
  }
}

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init);

  if (!response.ok) {
    throw await parseApiError(response);
  }

  return response.json() as Promise<T>;
}

export function fetchDeclaredSummary(
  analysisId: string,
  year: string,
): Promise<DeclaredLayerSummary> {
  return fetchJson<DeclaredLayerSummary>(
    `/api/v1/analyses/${analysisId}/exercises/${year}/declared`,
  );
}

export function fetchDeclaredAccounts(
  analysisId: string,
  year: string,
): Promise<DeclaredAccountsResponse> {
  return fetchJson<DeclaredAccountsResponse>(
    `/api/v1/analyses/${analysisId}/exercises/${year}/declared/accounts`,
  );
}

export function fetchDeclaredBalanceAccounts(
  analysisId: string,
  year: string,
): Promise<DeclaredBalanceResponse> {
  return fetchJson<DeclaredBalanceResponse>(
    `/api/v1/analyses/${analysisId}/exercises/${year}/declared/balance/accounts`,
  );
}

export function fetchDeclaredBalanceComponents(
  analysisId: string,
  year: string,
  aggregationCode: string,
): Promise<DeclaredBalanceComponentsResponse> {
  return fetchJson<DeclaredBalanceComponentsResponse>(
    `/api/v1/analyses/${analysisId}/exercises/${year}/declared/balance/accounts/${encodeURIComponent(aggregationCode)}/components`,
  );
}

export function importEcd(file: File): Promise<EcdImportResponse> {
  const formData = new FormData();
  formData.append("file", file);

  return fetchJson<EcdImportResponse>("/api/v1/ecd/import", {
    body: formData,
    method: "POST",
  });
}

export function fetchEcdImports(): Promise<EcdImportListResponse> {
  return fetchJson<EcdImportListResponse>("/api/v1/ecd/imports");
}

export function deleteEcdImport(ecdFileId: string): Promise<EcdImportDeleteResponse> {
  return fetchJson<EcdImportDeleteResponse>(`/api/v1/ecd/imports/${ecdFileId}`, {
    method: "DELETE",
  });
}

export function runDeclaredLayer(
  analysisId: string,
  year: number,
): Promise<DeclaredRunResponse> {
  return fetchJson<DeclaredRunResponse>(
    `/api/v1/analyses/${analysisId}/exercises/${year}/declared/run`,
    { method: "POST" },
  );
}
